src/Processors/QueryPlan/ReadFromRemote.cpp

```diff
@@ -472,12 +494,33 @@ void ReadFromRemote::addLazyPipe(Pipes & pipes, const ClusterProxy::SelectStream
                 max_remote_delay = std::max(try_result.delay, max_remote_delay);
         }
 
-        if (try_results.empty() || local_delay < max_remote_delay)
+        const UInt64 local_delay = replicated_storage->getAbsoluteDelay();
+        const UInt64 max_allowed_delay = current_settings[Setting::max_replica_delay_for_distributed_queries];
+
+        bool use_delayed_remote_source = false;
+        fiu_do_on(FailPoints::use_delayed_remote_source,
+        {
+            use_delayed_remote_source = true;
+        });
+
+        if (!use_delayed_remote_source)
         {
-            auto plan = createLocalPlan(
-                query, header, my_context, my_stage, my_shard.shard_info.shard_num, my_shard_count, my_shard.has_missing_objects);
+            if (try_results.empty() && local_delay >= max_allowed_delay)
```

Comment:
Can you please add a comment why this could not happens under normal circumstances? I.e. why it is `LOGICAL_ERROR` here

Assume that in `SelectStreamFactory::createForShardImpl`:
- `prefer_localhost_replica=1`
- `max_replica_delay_for_distributed_queries=5`
- `local_delay=10`

And other replica is unavailable, so `try_result.empty() == true` (_am I right that this is the case for this?_), so it will enter here.

What did I miss?

Reply:
Indeed, you are right: we don't add entries for connections that we failed to establish to `try_results`. Changed the error to `REPLICA_IS_STALE`.